File: src/pricing.ts
import { invoke } from "@tauri-apps/api/core";
import type { TokenUsage } from "./tauri";

function isTauriRuntimeForPricing(): boolean {
  return "__TAURI_INTERNALS__" in window;
}

export async function getPricing(): Promise<PricingResponse> {
  if (isTauriRuntimeForPricing()) {
    return invoke<PricingResponse>("get_pricing");
  }
  return { models: [], catalogFetchedAt: null, lastRefreshError: null };
}

export async function setModelRate(request: SetModelRateRequest): Promise<PricingResponse> {
  if (isTauriRuntimeForPricing()) {
    return invoke<PricingResponse>("set_model_rate", { request });
  }
  return { models: [], catalogFetchedAt: null, lastRefreshError: null };
}

export async function resetModelRate(modelId: string): Promise<PricingResponse> {
  if (isTauriRuntimeForPricing()) {
    return invoke<PricingResponse>("reset_model_rate", { modelId });
  }
  return { models: [], catalogFetchedAt: null, lastRefreshError: null };
}

export async function hideModel(modelId: string): Promise<PricingResponse> {
  if (isTauriRuntimeForPricing()) {
    return invoke<PricingResponse>("hide_model", { modelId });
  }
  return { models: [], catalogFetchedAt: null, lastRefreshError: null };
}

export async function refreshPricing(): Promise<PricingResponse> {
  if (isTauriRuntimeForPricing()) {
    return invoke<PricingResponse>("refresh_pricing");
  }
  return { models: [], catalogFetchedAt: null, lastRefreshError: null };
}

export type { ModelPricingEntry, ModelRate, PricingResponse, SetModelRateRequest };

export interface SetModelRateRequest {
  modelId: string;
  displayName?: string;
  rate: ModelRate;
}


export interface ModelRate {
  inputPerMillion: number;
  outputPerMillion: number;
  cacheReadPerMillion: number;
  cacheWritePerMillion: number;
}

export interface ModelPricingEntry {
  modelId: string;
  displayName: string;
  rate: ModelRate;
  isCustom: boolean;
  isUnpriced: boolean;
}

export interface PricingResponse {
  models: ModelPricingEntry[];
  catalogFetchedAt: string | null;
  lastRefreshError: string | null;
}

export const UNKNOWN_MODEL = "_unknown";

export function usageCostUsd(usage: TokenUsage, rate: ModelRate): number {
  return (
    (usage.inputTokens * rate.inputPerMillion +
      usage.outputTokens * rate.outputPerMillion +
      usage.cacheReadTokens * rate.cacheReadPerMillion +
      usage.cacheCreationTokens * rate.cacheWritePerMillion) /
    1_000_000
  );
}

export function pricingRateMap(models: ModelPricingEntry[]): Record<string, ModelRate> {
  return Object.fromEntries(
    models.filter((model) => !model.isUnpriced).map((model) => [model.modelId, model.rate]),
  );
}

export function byModelCostUsd(
  byModel: Record<string, TokenUsage> | undefined,
  rates: Record<string, ModelRate>,
): number {
  if (!byModel) return 0;
  let total = 0;
  for (const [modelId, usage] of Object.entries(byModel)) {
    if (modelId === UNKNOWN_MODEL) continue;
    const rate = rates[modelId];
    if (!rate) continue;
    total += usageCostUsd(usage, rate);
  }
  return total;
}

File: src-tauri/src/commands/pricing.rs
// Pricing catalog and model-rate commands.
use crate::*;

#[tauri::command]
pub(crate) fn get_token_history(days: u32) -> Result<token_history::TokenHistoryResponse, String> {
    token_history::get_token_history(days)
}

#[tauri::command]
pub(crate) fn get_pricing() -> Result<pricing::PricingResponse, String> {
    pricing::get_pricing()
}

#[tauri::command]
pub(crate) fn set_model_rate(
    request: pricing::SetModelRateRequest,
) -> Result<pricing::PricingResponse, String> {
    pricing::set_model_rate(request)
}

#[tauri::command]
pub(crate) fn reset_model_rate(model_id: String) -> Result<pricing::PricingResponse, String> {
    pricing::reset_model_rate(model_id)
}

#[tauri::command]
pub(crate) fn hide_model(model_id: String) -> Result<pricing::PricingResponse, String> {
    pricing::hide_model(model_id)
}

#[tauri::command]
pub(crate) fn unhide_model(model_id: String) -> Result<pricing::PricingResponse, String> {
    pricing::unhide_model(model_id)
}

#[tauri::command]
pub(crate) async fn refresh_pricing() -> Result<pricing::PricingResponse, String> {
    tauri::async_runtime::spawn_blocking(|| pricing::refresh_pricing_catalog(true))
        .await
        .map_err(|e| e.to_string())?
}
